### PR TITLE
Todos app - Marjaana

### PR DESCRIPTION
**Pull request by MayaBlixt**
Link to my deployed project:
 https://marjaanatodoapp.netlify.app/

My reflections on how this week's project turned out:

Things I'd like to have clarified or explained in more detail:

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,25 @@
...
 
-Describe how you approached to problem, and what tools and techniques you used to solve it. How did you plan? What technologies did you use? If you had more time, what would be next?
+## Limitations
+
+All blue level requirements achived: 
+- Your app should list all todo tasks - completed or uncompleted
+- You should be able to mark an uncompleted task as complete (and change it back to uncomplete)
+- You should be able to add and remove tasks
+- Your app should show a count of either all todos, or all uncomplete todos (or both)
+
```
Comment (luuf81) [inline README.md:19]: Very clear and comprehensive readme, great work!

File: README.md [hunk 2]
```diff
@@ -1,13 +1,25 @@
...
-Replace this readme with your own information about your project.
 
-Start by briefly describing the assignment in a sentence or two. Keep it short and to the point.
+## Accessibility and cross-browser testing 
 
-## The problem
+Html validated with W3C validation service (http://validator.w3.org/) with no errors found.
+Chrome lighthouse report was generated with the result: 78% Accessibility, 100% Best Practices and 100% SEO. 
+Accessibility checked with Wave chrome plugin and no errors raised and with screenreader chromeVox with no errors.
+Testing has been done in different browsers such as Edge, FireFox and Chrome in mobile, tablet and desktop views.
```
Comment (luuf81) [inline README.md:10]: You put effort on checking accesibility 👍

File: code/src/App.js [hunk 3]
```diff
@@ -1,9 +1,22 @@
-import React from 'react'
+import React from 'react';
+import { Provider } from 'react-redux';
+import { combineReducers, configureStore } from '@reduxjs/toolkit';
+
+import { todos } from 'reducers/todos';
+import  List  from 'components/List';
+import AddTodo from 'components/AddTodo';
+import { MainContainer } from 'styled-components/MainContainer';
```
Comment (luuf81) [inline code/src/App.js:8]: separating external from internal imports, good practice!

File: code/src/styled-components/AddTodo.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import styled from 'styled-components';
```
Comment (luuf81) [inline code/src/styled-components/AddTodo.js:1]: I like the fact that you used styled components in this project, that's what I also like to do. Only thing that's a bit confusing is that name of this file is AddTodo.js but there's only styling in here?

File: code/src/reducers/todos.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import { createSlice } from  '@reduxjs/toolkit';
+
+const initialState = {
+    items:[
+       { id:'1', name:'A first todo', isComplete: false},
+
```
Comment (luuf81) [inline code/src/reducers/todos.js:6]: redundant extra line here

File: code/src/reducers/todos.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+            const filteredList = store.items.filter(item => item.id !== itemId);
+            store.items = filteredList;
+            
+
+        },
+        addItem: (store, action) => {
+            const newItem = {
+                id: Math.max(...store.items.map(item =>item.id)) + 1,
+                name: action.payload,
+                isComplete: false
```
Comment (luuf81) [inline code/src/reducers/todos.js:25]: Great way of adding new ids, I used similar approach :)

File: code/src/components/List.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    return (
+        <>
+            <ListContainer>
+                {items.map(item => (
+                    <Item
+                        key={item.id}
+                        id={item.id}
+                        name={item.name}
+                        isComplete={item.isComplete}
+                    />
```
Comment (luuf81) [inline code/src/components/List.js:23]: It seems there margins are a bit of here, the list items doesn't seem to be aligned (in the app that is)

File: code/src/components/Item.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    }
+    const onIsCompletedChange = () => {
+        dispatch(todos.actions.toggleComplete(id));
+    };
+
+    return (
+        <TodoItem>
+            <div>{name}</div>
+            <ClickedButton onClick={onIsCompletedChange}>
+                {isComplete ? "☒" : "☑︎"}
```
Comment (luuf81) [inline code/src/components/Item.js:21]: Clean way of solving the check/uncheck requirement 👍

File: code/src/App.js [hunk 9]
```diff
@@ -1,9 +1,22 @@
-import React from 'react'
+import React from 'react';
+import { Provider } from 'react-redux';
+import { combineReducers, configureStore } from '@reduxjs/toolkit';
+
+import { todos } from 'reducers/todos';
+import  List  from 'components/List';
+import AddTodo from 'components/AddTodo';
```
Comment (luuf81) [inline code/src/App.js:7]: Curious as why you are importing these components without curly brackets?

File: code/src/styled-components/MainContainer.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import styled from 'styled-components';
+
+export const MainContainer = styled.main`
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+    justify-content: center;
```
Comment (luuf81) [inline code/src/styled-components/MainContainer.js:7]: tip: there's a shorthand for justify-content & align-items (if value is the same) -  "place-content"

File: netlify.toml [hunk 11]
```diff
@@ -3,4 +3,4 @@
 [build]
   base = "code/"
   publish = "build/"
-  command = "npm run build"
+  command = "CI= npm run build"
```
Comment (luuf81) [inline netlify.toml:6]: Used in most of my projects as well 😍

**Review by luuf81 — APPROVED** [review]
Great work on the ToDo project Marjaana! 

Above all I liked that the code was easy to read, you had straight forward and clean solutions for all the calculations, array alterations etc.

I also liked that you used styled-components extensively for the styling part of the app.

For me you have clearly met all the requirements on blue level and you also did styled-components which is red level, great work and Approved from my side 💪.

File: code/src/components/Item.js (hunk 8 above)
Comment (HIPPIEKICK) [reply code/src/components/Item.js:21]: I'm a little bit confused on which is which of these "☒" : "☑︎" 🙈

File: code/src/App.js (hunk 9 above)
Comment (HIPPIEKICK) [reply code/src/App.js:7]: Olof, here you can read about the difference between named import and default import 👇https://stackoverflow.com/a/36796281/12131107

File: code/src/styled-components/MainContainer.js (hunk 10 above)
Comment (HIPPIEKICK) [reply code/src/styled-components/MainContainer.js:7]: I learned something new, thanks Olof!

File: code/src/App.js [hunk 12]
```diff
@@ -1,9 +1,22 @@
...
 
 export const App = () => {
   return (
-    <div>
-      Find me in src/app.js!
-    </div>
+    <Provider store={store}> 
+    <MainContainer tabIndex="0">
+      <AddTodo></AddTodo>
+      <List></List>
```
Comment (HIPPIEKICK) [inline code/src/App.js:18]: These can be self closing

File: code/src/App.js [hunk 13]
```diff
@@ -1,9 +1,22 @@
...
+      </MainContainer>
+    </Provider>
   )
 }
```
Comment (HIPPIEKICK) [inline code/src/App.js]: So, semi colons or no semi colons? ;)

File: code/src/components/AddTodo.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
+import React, { useState } from 'react';
+import { useDispatch, useSelector } from 'react-redux'
```
Comment (HIPPIEKICK) [inline code/src/components/AddTodo.js]: Semi colons or no semi colons ;)

File: code/src/components/AddTodo.js [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+import { todos } from '../reducers/todos';
+import { InputContainer,InputNrOfTask, Input, InputButton, InputHeading, SetInputContainer } from '../styled-components/AddTodo';
+
+const AddTodo = () => {
+    const [value, setValue]= useState ('');
+    const dispatch = useDispatch();
+    const items = useSelector(store => store.todos.items);
+
+    const OnItemAdd = () =>{
+        dispatch(todos.actions.addItem(value))
```
Comment (HIPPIEKICK) [inline code/src/components/AddTodo.js]: ;

File: code/src/components/Item.js [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+            </ClickedButton>
+            <RemoveButton onClick={itemDelete}>x</RemoveButton>   
+        </TodoItem>
+    );
+};
+    export default Item;
```
Comment (HIPPIEKICK) [inline code/src/components/Item.js]: Remove indent

File: code/src/components/AddTodo.js [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+        <InputHeading> Todo list </InputHeading>
+        <InputNrOfTask>You have added {items.length}{' '}{items.length ===1 ? 'task' : 'tasks' }</InputNrOfTask>
+            <Input
+                id="task"
+                type= "text" 
+                value={value} 
+                onChange={e =>setValue(e.target.value)}
+                placeholder="Add todo" 
+            />
+        <InputButton onClick={OnItemAdd} type="submit" disabled={value.length < 1}> + </InputButton>
```
Comment (HIPPIEKICK) [inline code/src/components/AddTodo.js:27]: Wouldn't it be nice to clear the input value onclick as well? 😇

**Comment by HIPPIEKICK** [comment]
Get your semi colons sorted out and this is approved 💯

**Comment by MayaBlixt** [comment]
Added semicolons and fixed indentation :)

File: code/src/components/AddTodo.js (hunk 17 above)
Comment (MayaBlixt) [reply code/src/components/AddTodo.js:27]: Good suggestion 👍 :)

File: code/src/App.js [hunk 18]
```diff
@@ -1,9 +1,22 @@
...
-    <div>
-      Find me in src/app.js!
-    </div>
+    <Provider store={store}> 
+    <MainContainer tabIndex="0">
+      <AddTodo></AddTodo>
+      <List></List>
+      </MainContainer>
+    </Provider>
   )
```
Comment (HIPPIEKICK) [inline code/src/App.js:21]: Semicolon ;)

File: code/src/components/AddTodo.js [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+import { InputContainer,InputNrOfTask, Input, InputButton, InputHeading, SetInputContainer } from '../styled-components/AddTodo';
+
+const AddTodo = () => {
+    const [value, setValue]= useState ('');
+    const dispatch = useDispatch();
+    const items = useSelector(store => store.todos.items);
+
+    const OnItemAdd = () =>{
+        dispatch(todos.actions.addItem(value));
+    }
```
Comment (HIPPIEKICK) [inline code/src/components/AddTodo.js:14]: ;

File: code/src/components/AddTodo.js [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+            <Input
+                id="task"
+                type= "text" 
+                value={value} 
+                onChange={e =>setValue(e.target.value)}
+                placeholder="Add todo" 
+            />
+        <InputButton onClick={OnItemAdd} type="submit" disabled={value.length < 1}> + </InputButton>
+        </InputContainer>
+    )
```
Comment (HIPPIEKICK) [inline code/src/components/AddTodo.js:29]: ;

File: code/src/components/Item.js [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+
+import { todos } from '../reducers/todos';
+import { TodoItem, RemoveButton, ClickedButton } from 'styled-components/Item';
+
+const Item = ({ id, name, isComplete}) => {
+    const dispatch = useDispatch();
+
+    const itemDelete = () => {
+        dispatch(todos.actions.removeItem(id));
+    }
```
Comment (HIPPIEKICK) [inline code/src/components/Item.js:12]: ;

**Pull request closed without merging**